module/zfs/dmu_recv.c

```diff
@@ -1246,6 +1247,49 @@ dmu_recv_begin(char *tofs, char *tosnap, dmu_replay_record_t *drr_begin,
 
 	uint32_t payloadlen = drc->drc_drr_begin->drr_payloadlen;
 	void *payload = NULL;
+
+	/*
+	 * If the value passed to `kmem_alloc()` is sufficiently large on
+	 * either a low memory non-Linux system or Linux system, we will hang
+	 * in kmem_alloc(). Illumos and FreeBSD do not define precise upper
+	 * limits for kmem_alloc(), but Linux does.
+	 *
+	 * The maximum allocation size on Linux varies based on architecture
+	 * and kernel allocator. However, the minimum maximum size is 2^22 or
+	 * 4MB on architectures with 4KB pages. We handle Linux systems by
+	 * comparing against MIN(2^22, arc_all_memory() / 4). The inclusion of
+	 * arc_all_memory() in the calculation is intended to handle edge cases
+	 * involving extremely low memory Linux systems.
+	 *
+	 * We handle non-Linux systems with small amounts of memory by
+	 * comparing against arc_all_memory() / 4 on 64-bit systems and against
+	 * MIN(1U << 30, arc_all_memory()) on 32-bit systems. This should be
+	 * the largest allocation that is reasonable even on a system that
+	 * allows you to allocate nearly all memory inside the kernel, since
+	 * nvlist_unpack() will effectively double usage and we cannot expect
+	 * all memory on the system to be freed to acommedate us.
+	 *
+	 * The reason we handle 32-bit systems differently than 64-bit systems
+	 * here is because the kernel address space is typically limited to
+	 * 2GB, so if 4GB of RAM is installed, we will risk doing a 1GB memory
+	 * allocation, that will require slightly more than 1GB of additional
+	 * memory unpack, which cannot be allocated when the kernel is limited
+	 * to 2GB of address space for itself. If the region used to satisfy
+	 * kmem_alloc() is even smaller, then this would still be too large, so
+	 * we further limit the amount of memory this calculation allows us to
+	 * allocate here to 256MB as a precaution.
+	 */
+#ifdef __linux__
+	if (payloadlen > MIN((1U << 22), arc_all_memory() / 4))
+		return (E2BIG);
+#elif defined(_ILP32)
+	if (payloadlen > (MIN((1U << 28), arc_all_memory() / 4)))
+		return (E2BIG);
+#else
+	if (payloadlen > (arc_all_memory() / 4))
+		return (E2BIG);
```

Comment:
There is also a hypothetical kernel/user split issue on 64-bit too, but that is far in the future, so I did not bother looking up how to handle that. Extended level page tables also would make it difficult to pick a number for the split unless we just assume 4-level page tables.